xml/tags-3.3.rng

```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<grammar xmlns="http://relaxng.org/ns/structure/1.0" 
+         datatypeLibrary="http://www.w3.org/2001/XMLSchema-datatypes">
+  <start>
+    <optional>
+      <ref name="element-tags"/>
+    </optional>
+  </start>
+
+  <define name="element-tags">
+    <element name="tags">
+      <zeroOrMore>
+        <element name="tag">
+          <attribute name="id"><data type="ID"/></attribute>
+          <oneOrMore>
+            <element name="obj_ref">
+              <attribute name="id"><text /></attribute>
```

Comment:
Bummer. Unfortunately the type for node IDs is intentional -- they come from the underlying cluster stack, and an integer value has to be allowed.

We have a handful of scheduler regression tests with tags, but they're all resource tags since the scheduler only uses them in resource constraints. Obviously no one's bothered trying to use tags with nodes before. The documentation doesn't explicitly mention nodes, but it does say "Pacemaker allows you to tag any configuration element that has an XML ID."

I'm thinking that anything that processes tags should already ignore references that aren't of the type it's interested in. E.g. anything allowing a tag in place of a resource ID has to ignore any references in the tag that point to something other than resources. So switching to text shouldn't open up too many possibilities for trouble. Now whether the existing code actually does check that is a separate question :)

In hindsight, the proper thing to do would have been to make the node "id" attribute a CIB-internal XML ID, and use a separate field for something like "cluster-id". We could theoretically switch to that at a 2.x.0 version (using an XSL transform to upgrade existing CIBs) but it might not be worth the inevitable headaches.

Reply:
I wasn't very clear in the commit message, but it's not possible to load an XML patch into the CIB that contains tags on nodes with the existing schema.  Everything failed to load with errors that the configuration was not valid for the schema.  Hence, the change in tags-3.3.xml changing from an IDREF to text.  I agree that it seems like no one has tried to use a tag on a node.  Lucky for me that was the first thing I tried to do with the follow up to the --include/--exclude patches.